frontend/src/lib/Arrow.ts

```diff
@@ -0,0 +1,288 @@
+  field_name: string
+  metadata: Record<string, any> | null
+  name: string | null
+  numpy_type: string
+  pandas_type: string
+}
+
+interface TableDimensions {
+  headerRows: number
+  headerColumns: number
+  dataRows: number
+  dataColumns: number
+  rows: number
+  columns: number
+}
+
+type TableCellType = "blank" | "index" | "columns" | "data"
+
+interface TableCell {
+  type: TableCellType
+  id?: string
+  classNames: string
+  content: string
+}
+
+export interface Styler {
+  uuid?: string
+  caption?: string
+  styles?: string
+  displayValues?: Arrow
+}
+
+export class Arrow {
+  private table: Table
+
+  private schema: Schema
+
+  private index: string[][]
+
+  private columns: string[][]
+
+  private data: any[][]
+
+  private styler?: Styler
+
+  constructor(dataBuffer: Uint8Array, styler?: IStyler) {
+    this.table = Table.from(dataBuffer)
+    this.schema = this.getSchema()
+    this.index = this.getIndex()
+    this.columns = this.getColumns()
+    this.data = this.getData()
+    this.styler = {
+      uuid: styler?.uuid || undefined,
+      caption: styler?.caption || undefined,
+      styles: styler?.styles || undefined,
+      displayValues: styler?.displayValues
+        ? new Arrow(styler.displayValues)
+        : undefined,
+    }
+  }
+
+  private getSchema(): Schema {
+    const schema = this.table.schema.metadata.get("pandas")
+    if (schema == null) {
+      throw new Error("Table schema is missing.")
+    }
+    return JSON.parse(schema)
+  }
+
+  private static getColumnData(column: Column): any[] {
+    return range(0, column.length).map(rowIndex => column.get(rowIndex))
+  }
+
+  private getIndex(): string[][] {
+    return this.schema.index_columns.map(field => {
+      const isRangeIndex = typeof field === "object" && field.kind === "range"
+      if (isRangeIndex) {
+        const { start, stop, step } = field as RangeIndex
+        return range(start, stop, step).map(String)
+      }
+      const column = this.table.getColumn(field as string)
+      return Arrow.getColumnData(column).map(String)
+    })
+  }
+
+  private getColumns(): string[][] {
+    const isMultiIndex = this.schema.column_indexes.length > 1
+    return unzip(
+      this.schema.columns
+        .map(column => column.field_name)
+        .filter(fieldName => !this.schema.index_columns.includes(fieldName))
+        .map(fieldName =>
+          isMultiIndex
+            ? JSON.parse(
+                fieldName
+                  .replace(/\(/g, "[")
+                  .replace(/\)/g, "]")
+                  .replace(/'/g, '"')
+              )
+            : [fieldName]
+        )
+    )
+  }
+
+  private getData(): any[][] {
+    const rows = this.table.length
+    const columns = this.columns.length > 0 ? this.columns[0].length : 0
+
+    return range(0, rows).map(rowIndex =>
+      range(0, columns).map(columnIndex =>
+        this.table
+          .getColumnAt(columnIndex)
+          ?.get(rowIndex)
+          ?.toString()
```

Comment:
Not important at all but I find the ? on the new line confusing when it's modifying the previous line. My initial thought was this was supposed to be a ternary statement.